cli/options.go

```diff
@@ -384,6 +385,15 @@ func withConvertWindowsPaths(options *ProjectOptions) func(*loader.Options) {
 	}
 }
 
+// withSeparator defines loader.Options separator used to define resource names
+func withSeparator(options *ProjectOptions) func(*loader.Options) {
+	return func(o *loader.Options) {
+		if utils.StringToBool(options.Environment["COMPOSE_COMPATIBILITY"]) {
+			o.Separator = loader.CompatibilitySeparator
+		}
```

Comment:
Please note that this only fixes https://github.com/docker/compose/issues/9618 when passing an environment variable. In the case of Docker Compose getting a `--compatibility` from the command line I'm not sure that will work  (didn't test though).

Given this: https://github.com/docker/compose/blob/v2/cmd/compose/compose.go#L162-L169

Looks like compose just evaluates the command line option after this code is already evaluated.
One solution would be checking the command line option before and setting the EnvVar before calling `ProjectFromOptions` in the Docker Compose repo.